controllers/addon_controller.go

```diff
@@ -173,71 +182,106 @@ func (r *AddonReconciler) execAddon(ctx context.Context, req reconcile.Request,
-			IsController: true,
-			OwnerType:    &addonmgrv1alpha1.Addon{},
-		})
-
-	generatedInformers = informers.NewSharedInformerFactory(r.generatedClient, time.Minute*30)
-
-	err := mgr.Add(manager.RunnableFunc(func(s <-chan struct{}) error {
-		generatedInformers.Start(s)
-		generatedInformers.WaitForCacheSync(s)
-		nsInformers.Start(s)
-		nsInformers.WaitForCacheSync(s)
-		<-s
-		return nil
-	}))
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
 	if err != nil {
-		log.Error(err, "Error adding InformerFactory to the Manager")
-		return err
+		return nil, err
+	}
+
+	nsInformers := dynamicinformer.NewFilteredDynamicSharedInformerFactory(r.dynClient, time.Minute*30, workflowDeployedNS, nil)
+	wfInf := nsInformers.ForResource(common.WorkflowGVR())
+	// Watch workflows created by addon only in addon-manager-system namespace
+	if err := c.Watch(&source.Informer{Informer: wfInf.Informer().(cache.Informer)}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &addonmgrv1alpha1.Addon{},
+	}, predicate.NewPredicateFuncs(r.workflowHasMatchingNamespace)); err != nil {
+		return nil, err
 	}
 
 	// Watch for changes to kubernetes Resources matching addon labels.
-	for _, resc := range resources {
-		gvk := resc.GetObjectKind().GroupVersionKind()
-		_, kind := gvk.ToAPIVersionAndKind()
-
-		gvr := schema.GroupVersionResource{
-			Group:    gvk.Group,
-			Version:  gvk.Version,
-			Resource: inflection.Plural(strings.ToLower(kind)),
-		}
+	if err := c.Watch(&source.Kind{Type: &addonmgrv1alpha1.Addon{}}, &handler.EnqueueRequestForObject{},
+		predicate.NewPredicateFuncs(r.filterAddons)); err != nil {
+		return nil, err
+	}
 
-		inf, err := generatedInformers.ForResource(gvr)
-		if err != nil {
-			return err
-		}
+	if err := c.Watch(&source.Kind{Type: &appsv1.Deployment{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &v1.Service{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &appsv1.DaemonSet{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &appsv1.ReplicaSet{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &appsv1.StatefulSet{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &batchv1.Job{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+func (r *AddonReconciler) filterAddons(obj client.Object) bool {
+	addon, ok := obj.(*addonmgrv1alpha1.Addon)
+	if !ok {
+		r.Log.Error(fmt.Errorf("unexpected object type in addon watch predicates"), "expected", "*addonmgrv1alpha1.Addon", "found", reflect.TypeOf(obj))
+		return false
+	}
+
+	if AddonPkg.IsDuplicate(addon, r.versionCache) {
```

Comment:
Why is this here now?

Reply:
@kevdowney we discussed this previously, following is the discussion
https://github.com/keikoproj/addon-manager/pull/96#discussion_r803175059